### DISCUSSION TITLE
Allow custom types mapping to be defined globally on init phase

### DISCUSSION THREAD

**Discussion by QuestionAndAnswer** [mikro-orm/mikro-orm#4042] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

I developed my custom type `MyType` and instead of specifying it on every property like

```typescript
...
@Property({ type: MyType, /* or customType: MyType */ })
prop: My;
...
``` 
I would like to be able to set such mapping somewhere in init.

I tried to do that via `discovery.getMappedType` callback redefinition, but it wouldn't pick up my custom type, it just set it to the json as far as I can tell.

**Describe the solution you'd like**
I'd like to have callback mapping function or a straight map object to provide, which mikro-orm can refer to to get correct types mapping. 

**Describe alternatives you've considered**
1. Alternative is a manual mapping layer, which I currently has.
2. Or manual setting type/customType fields in decorators. This approach kinda defeats the purpose of TSMorph, as I have to specify then such things manually, instead of giving the framework the power to pick up types deterministically without errors.

**Additional context**
My real `MyType` implementation underlying storage is json compatible, so I don't have realy db level custom type. Maybe it's a bug instead of a feature, but I can't tell at the moment.

+ Implementing custom type I noticed that it is not handles arrays automatically, so if I have MyType[] definition somewhere, I have to process it also in MyType.

```typescript
export class MyType extends Type<
  MyData | MyData[] | null | undefined,
  string | null | undefined
> {
  public convertToDatabaseValue(
    value: MyData | MyData[] | null | undefined,
    platform: Platform
  ): string | null | undefined {
    let toSave: MyDataLike | MyDataLike[] | null = null;

    if (Array.isArray(value) && value.every((x) => x instanceof MyData)) {
      toSave = value.map((x) => x.toObject());
    }

    if (value instanceof MyData) {
      toSave = value.toObject();
    }

    if (platform.convertsJsonAutomatically(true) || !toSave) {
      return toSave as unknown as string;
    }

    return JSON.stringify(toSave);
  }

  public convertToJSValue(
    value: string | null | undefined,
    platform: Platform
  ): MyData | MyData[] | null | undefined {
    if (!value) return value as null | undefined;

    let toRestore: MyDataLike | MyDataLike[];

    if (!platform.convertsJsonAutomatically()) {
      toRestore = JSON.parse(value);
    } else {
      toRestore = value as unknown as MyDataLike | MyDataLike[];
    }

    return Array.isArray(toRestore)
      ? toRestore.map((x) => MyData.from(x))
      : MyData.from(toRestore);
  }

  public getColumnType(_: EntityProperty, platform: Platform): string {
    return platform.getJsonDeclarationSQL();
  }
}
```

**Comment by B4nan** [comment, marked as answer]
I think `getMappedType` should be the way to go, but checking the code now, I can see it is not used for the runtime type mapping, only for inferring the column type. Let me do some tests to see how performant it will be if we use the mapped types for everything internally, most of them as just blanks, passing the value from param to return value.

If it would be problematic perf-wise, we can have a second callback I guess.

Note that you should extend the ORM `JsonType` here, otherwise it might not be correctly handled, e.g. when you want to store a scalar or array to the JSON property.

**Reply by QuestionAndAnswer** [reply]
Regarding performance, not sure what you'd like to check, as I can set my type into decorator and kill the performance right now if I'd want to). I mean, that users already has a possibility to define runtime types mapping (manually for now, on each field). Having it in a single place for mapping, is having to do the same amount of work, just in a different place. But, that's IMHO. Of course you know internals better.

For the type extension hint - thanks!

**Reply by B4nan** [reply]
I mean the `getMappedType` does not do what you want, and if it will, it will affect every property type, not just the ones you would set this way, so it would affect everything. I have a benchmark to test this.

**Reply by QuestionAndAnswer** [reply]
Got it. Thanks.

**Reply by B4nan** [reply]
It has a rather small, but measurable effect. It will be mostly about the compiled hydration functions as they all become more complex and use outside binding for this. Will think of a way to do this only for the user overrides and not internal types.

**Reply by B4nan** [reply]
Here you go: 2f682e33587bf31c5020d66b1d963d60cbc39cac

**Reply by QuestionAndAnswer** [reply]
Thanks! So quick.
